src/ApparentHorizons/YlmSpherepack.cpp

```diff
@@ -525,44 +526,66 @@ YlmSpherepack::first_and_second_derivative(
   return result;
 }
 
-YlmSpherepack::InterpolationInfoPerPoint::InterpolationInfoPerPoint(
+template <typename T>
+YlmSpherepack::InterpolationInfo<T>::InterpolationInfo(
     const size_t m_max, const std::vector<double>& pmm,
-    const std::array<double, 2>& target)
-    : cos_theta(cos(target[0])),
+    const std::array<T, 2>& target_points)
+    : cos_theta(cos(target_points[0])),
       cos_m_phi(m_max + 1),
       sin_m_phi(m_max + 1),
-      pbar_factor(m_max + 1) {
-  const double sin_theta = sin(target[0]);
-  const double phi = target[1];
+      pbar_factor(m_max + 1),
+      num_points(get_size(target_points[0])) {
+
+  const auto& theta = target_points[0];
+  const auto& phi = target_points[1];
+
+  ASSERT(get_size(theta) == get_size(phi),
+         "Size mismatch for thetas and phis: " << get_size(theta) << " and "
+                                               << get_size(phi));
+  if (m_max < 2) {
+    ERROR("Must use m_max>=2, not m_max=" << m_max);
+  }
+
+  // `DataVectors` for working. `pbar_factor` is guaranteed to be at least size
+  // 3 as demanded by the `YlmSpherePack` constructor
+  T sinmphi{};
+  T cosmphi{};
+  auto& alpha = pbar_factor.at(0);
+  auto& beta = pbar_factor.at(1);
+  auto& deltasinmphi = pbar_factor.at(2);
 
   // Evaluate cos(m*phi) and sin(m*phi) by numerical recipes eq. 5.5.6
   {
-    cos_m_phi[0] = 1.0;
-    sin_m_phi[0] = 0.0;
-    const double beta = sin(phi);
-    const double alpha = 2.0 * square(sin(0.5 * phi));
-    double sinmphi = 0.0;
-    double cosmphi = 1.0;
+    alpha = 2.0 * square(sin(0.5 * phi));
+    beta = sin(phi);
+    sinmphi = make_with_value<T>(num_points, 0.);
+    cosmphi = make_with_value<T>(num_points, 1.);
+    cos_m_phi[0] = make_with_value<T>(num_points, 1.);
+    sin_m_phi[0] = make_with_value<T>(num_points, 0.);
     for (size_t m = 1; m < m_max + 1; ++m) {
-      const double deltacosmphi = alpha * cosmphi + beta * sinmphi;
-      const double deltasinmphi = alpha * sinmphi - beta * cosmphi;
-      cosmphi -= deltacosmphi;
+      deltasinmphi = alpha * sinmphi - beta * cosmphi;
+      cosmphi -= alpha * cosmphi + beta * sinmphi;
       sinmphi -= deltasinmphi;
       cos_m_phi[m] = cosmphi;
       sin_m_phi[m] = sinmphi;
```

Comment:
These lines each do an allocation as well, right?  (Because cos_m_phi[m] and sin_m_phi[m] are initially default-constructed.)  Should cos_m_phi, sin_m_phi, and pbar_factor be part of a Variables (or non-temporary TmpBuffer) that is stored?

Reply:
Dont these objects need to know their size at compile time?

Reply:
Oh, "these objects" = "Variables and TmpBuffer".  Yes, Variables and TmpBuffer need to know the number of components at compile time, but m_max is known only at run time.  Would be nice to avoid having m_max allocations on each call of InterpolationInfo, but Variables and TmpBuffer won't do it.  There should be a way to allocate a block of memory of a runtime-decided size and point non-owning DataVectors into that block of memory, in the same way that Variables does.

Reply:
pointing into a `std::vector` is an option if the allocation cannot be reused. Something to consider

Reply:
yes, sorry, I should express myself more clearly. I am not sure what you mean @nilsdeppe, or what that would look like.  All mechanisms I can think of would require the mechanism Mark is referring to.

Reply:
If `InterpolationInfo` stored, in addition to the `std::vector<T>` a `std::vector<double>` that gets used as a buffer when you have `T` being a `DataVector` then you could do something like:

```cpp
if constexpr (std::is_same_v<T, DataVector>) {
  buffer_for_datavectors.resize(2 * (m_max + 1) * num_points);
  for (size_t m = 0; m < cos_m_phi.size(); ++m) {
    cos_m_phi[m].set_data_ref(&buffer_for_datavectors[m * num_pts], num_pts);
    sin_m_phi[m].set_data_ref(&buffer_for_datavectors[m * num_pts + (m_max + 1) * num_pts], num_pts);
  }
}
```
You could also combine the `pbar_factor` allocation in this way, which would mean you do get the full benefits of vectorization with `DataVector` _without_ having to pay the cost of the memory allocations. It's not the prettiest, and would probably be good to add a comment, but that's the most efficient way of handling this. Note, however, that the copy semantics will no longer be trivial because on copy you need to point the `DataVector`s (which you don't want to copy) into the new buffer (which you do copy). So it does add some complexity, you could wrap the `set_data_ref` business in a member function of `InterpolationInfo` and then just call that in the copy semantics.

Reply:
ah, I see, that should work, thank you. I started implementing it but now find it would require implementation of a swap function, the Rule of 5 and a unit test at which point `InterpolationInfo` should probably have its own hpp and cpp file. 

At this stage it seems to me that we should deal with this only if profiling shows that these allocations are performance critical.

Reply:
Memory allocations are the only thing we've found to be very predictably slow. If the shape map will only be invoked every ~100 time steps it probably won't matter, but if we are talking every time step then it'll be an issue, unless we restrict to l<5... A year and a half or so back I removed ~36 allocations per time step and got a 25% reduction in runtime, all work being equal. Here we're doing something like 2 allocations per `m`, so in order to not add those 36 allocations per time step back, we need to be at like l=4 or so at most... Allocations are _extremely_ expensive in a parallel/threaded environment because each allocation needs to go through the OS (effectively) and that means you have pretty close to a serial bottleneck. So depending on the use-case, I'm not really happy about the idea of deferring a massive reduction in allocations, partly because this will mean someone else (in the past often me) has to go through, understand the code, update, and test to make sure nothing broke.

Reply:
The current hope is to cache the `InterpolationInfo` in the shape map which every element holds an instance of, so that it would only have to be re-computed if the logical grid changes or an element gets split/fused, i.e. p- and h-refinement. 

But if it is that likely to be a bottleneck I can implement it. This could be a class in `DataStructures` which is templated on two containers, nested inside each other, which dynamically allocates them in the constructor as you described and properly takes care of copies and so on. The inner container would have to inherit from `VectorImpl` to be able to set the data reference. This can then be made iterable and return the non-owning inner containers. It can also be specialized on doubles to avoid the `if constexpr`.